ultravox/model/ultravox_pipeline.py

```diff
@@ -0,0 +1,116 @@
+        **kwargs
+    ):
+        if tokenizer is None:
+            tokenizer = transformers.AutoTokenizer.from_pretrained(
+                model.config._name_or_path
+            )
+
+        if audio_processor is None:
+            audio_processor = transformers.AutoProcessor.from_pretrained(
+                model.config.audio_model_id or model.config.audio_config._name_or_path
+            )
+
+        self.processor = UltravoxProcessor(
+            audio_processor=audio_processor,
+            tokenizer=tokenizer,
+            stack_factor=model.config.stack_factor,
+        )
+
+        super().__init__(model=model, tokenizer=tokenizer, **kwargs)
+
+    def _sanitize_parameters(self, **kwargs):
+        generation_kwargs = {}
+        if "temperature" in kwargs:
+            generation_kwargs["temperature"] = kwargs["temperature"]
+        if "max_new_tokens" in kwargs:
+            generation_kwargs["max_new_tokens"] = kwargs["max_new_tokens"]
+        if "repetition_penalty" in kwargs:
+            generation_kwargs["repetition_penalty"] = kwargs["repetition_penalty"]
+        return {}, generation_kwargs, {}
+
+    def preprocess(self, inputs: Dict[str, Any]):
+        if "turns" in inputs:
+            turns = inputs["turns"]
+        else:
+            prompt = inputs.get("prompt", "<|audio|>")
+            if "<|audio|>" not in prompt:
+                logging.warning(
+                    "Prompt does not contain '<|audio|>', appending '<|audio|>' to the end of the prompt."
+                )
+                prompt += " <|audio|>"
+            turns = [{"role": "user", "content": prompt}]
+
+        text = self.processor.tokenizer.apply_chat_template(turns, tokenize=False)
+
+        # TODO: allow text-only mode?
+        assert "audio" in inputs, "Audio input is required"
+
+        if "sampling_rate" not in inputs:
+            logging.warning(
+                "No sampling rate provided, using default of 16kHz. We highly recommend providing the correct sampling rate."
+            )
+
+        output = self.processor(
+            text=text,
+            audio=inputs["audio"],
+            sampling_rate=inputs.get("sampling_rate", 16000),
+        )
+        if "audio_values" in output:
+            output["audio_values"] = output["audio_values"].to(self.model.dtype)
+
+        return output
+
+    def _forward(
+        self,
+        model_inputs: Dict[str, Any],
+        temperature: Optional[float] = None,
+        max_new_tokens: Optional[int] = None,
+        repetition_penalty: float = 1.1,
+    ) -> List[int]:
+        temperature = temperature or None
+        do_sample = temperature is not None
+
+        terminators = [self.tokenizer.eos_token_id]
+        if "<|eot_id|>" in self.tokenizer.added_tokens_encoder:
+            terminators.append(self.tokenizer.convert_tokens_to_ids("<|eot_id|>"))
+
+        input_len = model_inputs["input_ids"].shape[1]
+
+        outputs = self.model.generate(
```

Comment:
I was thinking of that, but wasn't sure if it would be compatible with the HF Pipeline API. I'll verify.

Reply:
Okay I don't think it makes sense to support streaming here for two reasons:

1. The original `text_generation` pipeline doesn't support it
2. The parent class doesn't seem to be able to handle streaming (i.e. a generator).

We can probably support it, but it's more work due to 2 and people probably don't even expect it because of 1.

![image](https://github.com/user-attachments/assets/d93521ca-07c0-42f3-950a-fe262d05d40c)